[Inductor] simplify CPP backend Tile2D code and support non-contiguous load/store

jgong5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98138
* #98135
* #97781
* #97644
* __->__ #97626
* #97625

Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

This PR also adds a functional support for transposed copy of bfloat16 data types. Better performance requires vectorized intrinsics implemented for at::vec::transpose_mxn. cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire 

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Commit:
[Inductor] simplify CPP backend Tile2D code

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code"




cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code"




cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code and support non-contiguous load/store"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

This PR also adds a functional support for transposed copy of bfloat16 data types. Better performance requires vectorized intrinsics implemented for at::vec::transpose_mxn. cc zhuhaozhe 

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code and support non-contiguous load/store"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

This PR also adds a functional support for transposed copy of bfloat16 data types. Better performance requires vectorized intrinsics implemented for at::vec::transpose_mxn. cc zhuhaozhe 

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code and support non-contiguous load/store"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

This PR also adds a functional support for transposed copy of bfloat16 data types. Better performance requires vectorized intrinsics implemented for at::vec::transpose_mxn. cc zhuhaozhe 

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -6769,7 +6769,7 @@ def fn(x1, x2):
                 compiled = compile_fx_inner(traced, [x1, x2])
                 assert same(fn(x1, x2)[0], compiled([x1, x2])[0], equal_nan=True)
-                assert metrics.generated_cpp_vec_kernel_count == 1
+                assert metrics.generated_cpp_vec_kernel_count == 2
```
Why did this change?

jgong5 replied:
This is because the tail kernel of the tile2d is also vectorized along the outer loop index (i0 below). Originally it is a scalar kernel. The non-contiguous load falls back to the scalar inner loop (tmp11 below) and other code is vectorized as much as possible. Therefore, we have both tile2d main kernel and tile2d tail kernel vectorized, which is counted as 2 now.

Generated tail kernel of tile2d below:
```c++
                #pragma GCC ivdep
                for(long i1=static_cast<long>(0); i1<static_cast<long>(10); i1+=static_cast<long>(1))
                {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>((16*i0) + (20*i1)));
                    auto tmp11 = ([&]() { __at_align__ float tmpbuf[16]; for (long i0_inner = 0; i0_inner < 16; i0_inner++) tmpbuf[i0_inner] = in_ptr1[static_cast<long>(i1 + (10*i0_inner) + (160*i0))]; return at::vec::Vectorized<float>::loadu(tmpbuf); })();
                    auto tmp1 = tmp0.abs();
                    auto tmp2 = tmp1.sin();
                    auto tmp3 = tmp2.neg();
                    auto tmp4 = tmp3 * tmp3;
                    auto tmp5 = decltype(tmp4)(1)/(decltype(tmp4)(1) + tmp4.neg().exp());
                    auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                    auto tmp7 = tmp6.cos();
                    auto tmp8 = tmp7.exp();
                    auto tmp9 = tmp8.sqrt();
                    auto tmp10 = tmp9 + tmp0;
                    auto tmp12 = tmp10 - tmp11;
                    auto tmp13 = tmp12 * tmp0;
                    auto tmp14 = tmp13 / tmp0;
                    auto tmp15 = tmp14 * tmp14;
                    auto tmp16 = tmp15 * tmp15;
                    auto tmp17 = tmp16 * tmp14;
                    auto tmp18 = tmp17 * tmp17;
                    auto tmp19 = tmp18.log();
                    auto tmp20 = tmp19.floor();
                    auto tmp21 = tmp20.ceil();
                    auto tmp22 = tmp21.trunc();
                    auto tmp23 = tmp22.lgamma();
                    auto tmp24 = tmp23.fmod(tmp11);
                    auto tmp25 = decltype(tmp24)::blendv(decltype(tmp24)(0), decltype(tmp24)(1), decltype(tmp24)(0) < tmp24);
                    auto tmp26 = decltype(tmp24)::blendv(decltype(tmp24)(0), decltype(tmp24)(1), tmp24 < decltype(tmp24)(0));
                    auto tmp27 = tmp25 - tmp26;
                    auto tmp28 = tmp27 + tmp11;
                    tmp28.store(out_ptr0 + static_cast<long>((16*i0) + (20*i1)));
                }
```

jansel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -2244,35 +2171,88 @@ def run(kernel):
+                        return []
+                    stride = self.stride_at(var, index)
+                    if stride == 1:
+                        contig_vars.add(int(var.name[1:]))
```
Variable names can sometimes be stuff like `indirect0` so I don't think converting vars to ints will work in general.  Suggest adding sort function based on something like:
```py
def sort_key(x):
   return (*[int(m) for m in re.findall(r"\d+", x)], x)
```

jgong5 replied:
I intended only to check the loop variables and get the indices from them. So I revised the logic to make sure they are loop variables. Please check.

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code and support non-contiguous load/store"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

This PR also adds a functional support for transposed copy of bfloat16 data types. Better performance requires vectorized intrinsics implemented for at::vec::transpose_mxn. cc zhuhaozhe 

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code and support non-contiguous load/store"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

This PR also adds a functional support for transposed copy of bfloat16 data types. Better performance requires vectorized intrinsics implemented for at::vec::transpose_mxn. cc zhuhaozhe 

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

Commit:
Update on "[Inductor] simplify CPP backend Tile2D code and support non-contiguous load/store"


Remove `CppTile2DTailKernel` and `CppTile2DKernelChecker` and reuse `CppVecKernel` and `CppVecKernelChecker` for them. Add vectorization with fallback for load/store in CppVecKernel for the non-contiguous load/store needed by `CppTile2DTailKernel`.

This PR also adds a functional support for transposed copy of bfloat16 data types. Better performance requires vectorized intrinsics implemented for at::vec::transpose_mxn. cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 

cc soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

jgong5 commented:
@pytorchbot merge
